member/src/main/scala/com/improving/app/member/domain/Member.scala

```diff
@@ -56,220 +68,409 @@ object Member extends StrictLogging {
+        case terminateMemberCommand: TerminateMember => terminateMember(state, terminateMemberCommand, command.replyTo)
+        case editMemberInfoCommand: EditMemberInfo => editMemberInfo(state,  editMemberInfoCommand, command.replyTo)
+        case getMemberInfoCommand: GetMemberInfo => getMemberInfo(state, getMemberInfoCommand, command.replyTo)
+        case _ => useErrorStatusReply(command.replyTo, s"Invalid Command ${command.request}")
       }
   }
 
-  def registerMember(
-      memberId: MemberId,
-      memberInfo: MemberInfo,
-      actingMember: MemberId,
-      replyTo: ActorRef[StatusReply[MemberResponse]]
-  ): ReplyEffect[MemberEvent, MemberState] =
-    MemberValidation.validateMemberInfo(memberInfo) match {
-      case Valid(memberInfo) =>
-        val event =
-          MemberRegistered(Some(memberId), Some(memberInfo), Some(actingMember), Some(Timestamp(Instant.now(clock))))
-        Effect.persist(event).thenReply(replyTo) { _ =>
-          val res = StatusReply.Success(MemberEventResponse(event))
-          logger.info(s"registerMember: $res")
-          res
+  //EventHandler
+  private val eventHandler: (MemberState, MemberEvent) => MemberState = { (state, event) =>
+    event match {
+      case memberRegisteredEvent: MemberRegistered =>
+        state match {
+          case x: DraftMemberState =>
+            x.copy(
+              requiredInfo = x.requiredInfo.copy(
+                contact = memberRegisteredEvent.memberInfo.get.contact,
+                handle = memberRegisteredEvent.memberInfo.get.handle,
+                avatarUrl = memberRegisteredEvent.memberInfo.get.avatarUrl,
+                firstName = memberRegisteredEvent.memberInfo.get.firstName,
+                lastName = memberRegisteredEvent.memberInfo.get.lastName,
+                tenant = memberRegisteredEvent.memberInfo.get.tenant
+              ),
+              optionalInfo = x.optionalInfo.copy(
+                notificationPreference = memberRegisteredEvent.memberInfo.get.notificationPreference,
+                organizationMembership = memberRegisteredEvent.memberInfo.get.organizationMembership
+              ),
+              meta = memberRegisteredEvent.meta.get
+            )
+          case _: RegisteredMemberState => state
+          case _: TerminatedMemberState => state
         }
-      case Invalid(errors) =>
-        Effect.reply(replyTo) {
-          StatusReply.Error(
-            s"Invalid Member Info: $memberInfo with errors: ${errors.map { _.errorMessage }.toList.mkString(",")}"
+
+      case memberActivatedEvent: MemberActivated =>
+        state match {
+          case DraftMemberState(requiredInfo, optionalInfo, _) => RegisteredMemberState(
+            info = createMemberInfoFromDraftState(requiredInfo, optionalInfo),
+            meta = memberActivatedEvent.meta.get
           )
+          case x: RegisteredMemberState =>
+            if (x.meta.memberStatus.isMemberStatusSuspended) {
```

Comment:
I'm leaning a little towards asking you to get rid of this check. On the one hand it duplicates logic from the command handler - an event handler that doesn't have to check `currentState` might be ideal here. On the other hand it helps developers see more easily what's going on logically on the domain level. But that is exactly my concern that make me want to ignore `currentState` here as much as possible - the implementation is not the domain, and shouldn't necessarily act as documentation for it. What do you think?

Reply:
Hmm.. this piece of code is definitely more of a safety net rather than necessary code. After all, if it got to this point, it's because all validation checks have already passed. Can this be assumed? Or do we have to be devil's advocate here and assume that there may be some weird states that make this needed? I believe this should be the determining factor of whether or not it should be kept 🤔

Reply:
I suggest to apply the type system to get compile-time guarantees that an invalid state can't be reached here.

Reply:
@miguelTaningcoYoppworks in terms of command/event handler separation of concerns, where event handlers are only supposed to duplicate state and command handlers are only supposed to do validation - I don't think there is any need to be devil's advocate.

There may be cases where we are worried about eventual consistency & updates out of order or something like that, but I don't think this is one of those cases. You can just return `x.copy` every time because the only other possible state is Active, so if 2 Activate commands come in you are just mutating state from Active -> Active.